README
ch-smpp

The ch-smpp library is an extremely efficient, scalable, and flexible Java
implementation of the Short Messaging Peer to Peer Protocol (SMPP). This library
supports version 3.3, 3.4, and most of 5.0.  This library uses
Java non-blocking (NIO) sockets to support thousands of binds using minimal
resources.

The library has been tested and certified with hundreds of mobile operators
and suppliers around the world.  It's effective at being flexible with SMPP
specifications that aren't truly compliant.  The goal of the internal PDU
parser is to be unbreakable.


Contributors
--------------------------------------------------------------------------------

Joe Lauer (twitter: http://twitter.com/jjlauer) OR
    Text "FOLLOW jjlauer" to 40404


Installation
--------------------------------------------------------------------------------

Binary builds are publicly available at the Cloudhopper Maven repo:
http://maven.cloudhopper.com/repos/third-party/cloudhopper

Source builds require a parent POM available in a project:
http://maven.cloudhopper.com/repos/third-party/cloudhopper/ch-maven-parent-public


Demo Code / Tutorials
--------------------------------------------------------------------------------

For examples on using this library as a client or server, please see:
   src/test/java/com/cloudhopper/smpp/demo

To run some of the samples, there is a Makefile to automate running some of the
Maven commands.  On Linux/Mac, you'd run:

   make run-server
   make run-client

On Windows, you'd run:
 
   nmake run-server
   nmake run-client


License
--------------------------------------------------------------------------------

Copyright (C) 2011 Twitter, Inc.

This work is licensed under the Apache License, Version 2.0. See LICENSE for
details.


What do we need help with?
--------------------------------------------------------------------------------

There are always additional non-compliant SMPP implementations that exist out
in the wild.  Here are a few ways you can help us:

 1. Find nuances with other vendors, create unit tests, and implement a
    workaround.
 2. Additional demos and simulators
 3. Additional utility classes for TLV types
 4. Additional real world DeliveryReceipt formats
 5. SSL support (see Netty, both client and server side, custom certs)

How does SMPP "windowing" effect performance?
--------------------------------------------------------------------------------

With windowSize=2 took 36155 ms to process 1000 requests
With windowSize=10 took 10033 ms to process 1000 requests
With windowSize=100 took 3248 ms to process 1000 requests

CHANGELOG.md
Cloudhopper by Twitter
============================

cloudhopper-smpp
----------------
## 6.0.1
 - Updated netty to 4.1.36.Final

## 6.0.0-netty4-beta-2
 - ch-commons-util version bumped from 6.0.1 to 6.0.2 to fix race condition bug
   in WindowFuture:
     https://github.com/twitter/cloudhopper-smpp/issues/61

## 6.0.0-netty4-beta-1
 - Pre-release of port to Netty 4. Includes community contributions and changes based
   on Trustin's review.
 - Netty dependency changed to 4.0.25.Final.


NOTE: Project moved to https://github.com/fizzed/cloudhopper-smpp

## 5.0.9 - 2015-11-24
 - Add support for missing pdus and tags
 - Allow unbound channel to respond to enquire_link PDU
 - Corrected name of DataCoding SmppConstant to IA5 (not GSM)
 - Updated value of DATA_CODING_GSM constant and made it deprecated
 - Add ESM_CLASS Message Mode constants
 - DefaultSmppServer should use IO executor passed to its constructor
 - Fix-up comments on the SMPP error code constants

## 5.0.8 - 2015-04-17
 - Fixed issue where rawErrorCode not set on DeliveryReceipt (khaing211)
 - Support for host address in SmppServerConfiguration (pgoergler)
 - Improved demo for echo server (dwilkie)
 - Tlv class supports equals and hashCode (skytreader)

## 5.0.7 - 2015-02-02
 - ch-commons-util version bumped from 6.0.1 to 6.0.2 to fix race condition bug 
   in WindowFuture:
     https://github.com/twitter/cloudhopper-smpp/issues/61
 - Netty dependency bumped from 3.9.0.Final to 3.9.6.Final
 
## 5.0.6 - 2014-04-02
 - Support for low-level PDU listener (supports advanced logging, sniffing, and
   discarding before normal processing). New methods overridable in
   DefaultSmppSessionHandler are firePduRecived()
 - Bug with doneDate null-check fixed in DeliveryReceipt.toShortMessage().
 - Delivery receipt intermediate constant has a flag of bit 4 not 5. Please note
   that esm_class does used bit 5.
 - Bumped pom parent to v1.5
 - Dependencies now have specific version rather than version range.
 - Updated docs with user contributed demos

## 5.0.5 - 2014-01-07
 - Changed bindTimeout and writeTimeout implementations to adhere to documented
   best practices:
     http://netty.io/3.9/api/org/jboss/netty/channel/ChannelFuture.html
	 http://netty.io/3.9/api/org/jboss/netty/handler/timeout/WriteTimeoutHandler.html
   This solves a NullPointerException being thrown in the case where we were
   calling ChannelFuture.getCause() before the read/write was complete (and thus null).
 - Netty dependency changed from 3.7.0.Final to 3.9.0.Final

## 5.0.4 - 2013-10-17
 - Tweaked delivery receipt helper class to be more compliant w/ SMPP 3.4
   specs by parsing err code as String rather than as an int. Previous int
   methods were retained for backwards compatability, but accessing the raw
   error code as a String is now possible.
   Thx to williamd1618 for pull request:
     https://github.com/twitter/cloudhopper-smpp/pull/42

## 5.0.3 - 2013-10-13
 - Netty dependency bumped from 3.5.8.Final to 3.7.0.Final
 - Added support for a "writeTimeout" to channel writes.
 - Added support for PDUEncoder to not include message_id in some edge cases.
   Thx to chadselph for pull request:
     https://github.com/twitter/cloudhopper-smpp/pull/31

## 5.0.2 - 2013-04-10
 - Added SSL support for servers and clients with unit and integration tests.

## 5.0.1 - 2013-03-09
 - Added support for cancel_sm and query_sm SMPP messages. Added unit and 
   integration tests.
 - Cleaned up intermittent unit test failures by added delays before session
   close.

## 5.0.0 - 2012-10-26
 - No major code changes, mostly project layout changes in prep of release to
   Maven Central Repository
 - Maven artifact group changed from cloudhopper to com.cloudhopper
 - Readme and changelog switched to markdown
 - SLF4J, Netty, and Logback dependencies switched to use Maven Central groupIds
 - Updated underlying maven license plugin - every java file had its
   license positioned after package declaration.

## 4.1.2 - 2011-10-26
 - Upgrade parent POM dependency from 1.5 to 1.6
 - Added support for OSGI in final jar

## 4.1.1 - 2011-08-18
 - Fixed bug with DataSM decoding and encoding (previous impl was not per
    SMPP 3.4 specs).

## 4.1.0 - 2011-07-26
 - Upgraded ch-commons-util dependency from 5.0 to 5.1.0
 - Upgraded ch-commons-gsm dependency from 2.0 to 2.1.0
 - Upgrade parent POM dependency from 1.0 to 1.5
 - Switched to new version format.
 - Added Version class to compile build info into jar
 - Added SmppInvalidArgumentException class
 - Fixed issue with short messages > than 255 bytes in length not triggering
    an exception to be thrown when attempting to setShortMessage(). The length
    of the byte array is now checked.

## 4.0 - 2011-05-09
 - Upgraded ch-commons-util dependency from 4.1 to 5.0
 - WARN: There are some minor source and binary backwards compatability issues
    that were unavoidable to add request expiry as a new feature.
 - Automatic request expiration is fully supported.  See the new 
    constructors for both DefaultSmppClient and DefaultSmppServer that include
    a "monitorExecutor" parameter to enable the feature.  Also, there are new
    several new options for an SmppServerConfiguration that will control
    the default settings for each server session (including enable monitoring).
       defaultWindowSize
       defaultWindowMonitorInterval
       defaultWindowWaitTimeout
       defaultRequestExpiryTimeout
       maxConnectionSize
 - Added support for counters in both sessions and server for tracking 
    various key metrics.  Counters are enabled on an SmppServer by default, but
    must be explicitly enabled on sessions.  See countersEnableD() option.
 - Added initial support for JMX management by including an SmppServerMXBean
    and an SmppSessionMXBean.  Servers can automatically have themselves and
    their sessions registered & unregistered, but clients will need to use
    explictly register/unregister.

src/main/java/com/cloudhopper/smpp/tlv/Tlv.java
package com.cloudhopper.smpp.tlv;

/*
 * #%L
 * ch-smpp
 * %%
 * Copyright (C) 2009 - 2015 Cloudhopper by Twitter
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import com.cloudhopper.commons.util.ByteArrayUtil;
import com.cloudhopper.commons.util.HexUtil;
import java.io.UnsupportedEncodingException;
import java.util.Arrays;

/**
 * Tag-Length-Value optional parameter in SMPP.
 * 
 * @author joelauer (twitter: @jjlauer or <a href="http://twitter.com/jjlauer" target=window>http://twitter.com/jjlauer</a>)
 */
public class Tlv {
    
    private final short tag;
    private final byte[] value;     // length is stored in array
    private String tagName;      // short description of this tag

    public Tlv(short tag, byte[] value) {
        this(tag, value, null);
    }

    /**
     * The tagName is only for logging/debugging by showing up in the toString() method.
     * @param tag
     * @param value
     * @param tagName
     */
    public Tlv(short tag, byte[] value, String tagName) {
        this.tag = tag;
        this.value = value;
        this.tagName = tagName;
    }

    public String getTagName() {
        return this.tagName;
    }

    public void setTagName(String value) {
        this.tagName = value;
    }

    public short getTag() {
        return this.tag;
    }

    /**
     * Gets an "unsigned" version of this TLV's length of its value.  A TLV
     * has a 16-bit integer that is encoded as a short.  Since Java only has
     * "signed" shorts, any length > 32K would end up being a negative value.
     * This method returns the "unsigned" version as an int.
     * @return The "unsigned" length of this TLV's value
     */
    public int getUnsignedLength() {
        return (value == null ? 0 : value.length);
    }

    public short getLength() {
        return (short)getUnsignedLength();
    }

    public byte[] getValue() {
        return this.value;
    }

    /**
     * Returns the size of this TLV in bytes.  Basically, its always 4 bytes
     * plus the length of the value.  Two bytes for tag and two byte for length
     * and then the length of the value itself.  Primarily used for encoding
     * the PDU.
     * @return
     */
    public int calculateByteSize() {
        return 4 + getUnsignedLength();
    }

    public String getValueAsString() throws TlvConvertException {
        return getValueAsString("ISO-8859-1");
    }

    public String getValueAsString(String charsetName) throws TlvConvertException {
        if (this.value == null) {
            return null;
        }
        if (this.value.length == 0) {
            return "";
        }
        // default the position to be the entire byte array
        int len = this.value.length;
        for (int i = 0; i < this.value.length; i++) {
            if (this.value[i] == 0x00) {
                len = i;
                break;
            }
        }

        try {
            return new String(this.value, 0, len, charsetName);
        } catch (UnsupportedEncodingException e) {
            throw new TlvConvertException("String", "unsupported charset " + e.getMessage());
        }
    }

    @Override
    public String toString() {
        StringBuilder buffer = new StringBuilder(40);
        if (tagName != null) {
            buffer.append(this.tagName);
        } else {
            buffer.append("tlv");
        }
        buffer.append(": 0x");
        buffer.append(HexUtil.toHexString(this.tag));
        buffer.append(" 0x");
        buffer.append(HexUtil.toHexString((short)getUnsignedLength()));
        buffer.append(" [");
        HexUtil.appendHexString(buffer, this.value);
        buffer.append("]");
        return buffer.toString();
    }

    /**
     * Attempts to get the underlying value as a byte.  If the underlying byte
     * array cannot be converted, then an exception is thrown.  For example,
     * if the underlying byte array isn't a length of 1, then it cannot be
     * converted to a byte.
     * @return The byte array value as a byte
     * @throws TlvConvertException Thrown if the underlying byte array cannot
     *      be converted to a byte.
     */
    public byte getValueAsByte() throws TlvConvertException {
        try {
            return ByteArrayUtil.toByte(this.value);
        } catch (IllegalArgumentException e) {
            throw new TlvConvertException("byte", e.getMessage());
        }
    }

    /**
     * Attempts to get the underlying value as an unsigned byte.  If the underlying byte
     * array cannot be converted, then an exception is thrown.  For example,
     * if the underlying byte array isn't a length of 1, then it cannot be
     * converted to an unsigned byte.
     * @return The byte array value as an unsigned byte
     * @throws TlvConvertException Thrown if the underlying byte array cannot
     *      be converted to an unsigned byte.
     */
    public short getValueAsUnsignedByte() throws TlvConvertException {
        try {
            return ByteArrayUtil.toUnsignedByte(this.value);
        } catch (IllegalArgumentException e) {
            throw new TlvConvertException("unsigned byte", e.getMessage());
        }
    }

    /**
     * Attempts to get the underlying value as a short.  If the underlying byte
     * array cannot be converted, then an exception is thrown.  For example,
     * if the underlying byte array isn't a length of 2, then it cannot be
     * converted to a short.
     * @return The byte array value as a short
     * @throws TlvConvertException Thrown if the underlying byte array cannot
     *      be converted to a short.
     */
    public short getValueAsShort() throws TlvConvertException {
        try {
            return ByteArrayUtil.toShort(this.value);
        } catch (IllegalArgumentException e) {
            throw new TlvConvertException("short", e.getMessage());
        }
    }

    /**
     * Attempts to get the underlying value as an unsigned short.  If the underlying byte
     * array cannot be converted, then an exception is thrown.  For example,
     * if the underlying byte array isn't a length of 2, then it cannot be
     * converted to an unsigned short.
     * @return The byte array value as an unsigned short
     * @throws TlvConvertException Thrown if the underlying byte array cannot
     *      be converted to an unsigned short.
     */
    public int getValueAsUnsignedShort() throws TlvConvertException {
        try {
            return ByteArrayUtil.toUnsignedShort(this.value);
        } catch (IllegalArgumentException e) {
            throw new TlvConvertException("unsigned short", e.getMessage());
        }
    }

    /**
     * Attempts to get the underlying value as an int.  If the underlying byte
     * array cannot be converted, then an exception is thrown.  For example,
     * if the underlying byte array isn't a length of 4, then it cannot be
     * converted to an int.
     * @return The byte array value as an int
     * @throws TlvConvertException Thrown if the underlying byte array cannot
     *      be converted to an int.
     */
    public int getValueAsInt() throws TlvConvertException {
        try {
            return ByteArrayUtil.toInt(this.value);
        } catch (IllegalArgumentException e) {
            throw new TlvConvertException("int", e.getMessage());
        }
    }

    /**
     * Attempts to get the underlying value as an unsigned int.  If the underlying byte
     * array cannot be converted, then an exception is thrown.  For example,
     * if the underlying byte array isn't a length of 4, then it cannot be
     * converted to an unsigned int.
     * @return The byte array value as an unsigned int
     * @throws TlvConvertException Thrown if the underlying byte array cannot
     *      be converted to an unsigned int.
     */
    public long getValueAsUnsignedInt() throws TlvConvertException {
        try {
            return ByteArrayUtil.toUnsignedInt(this.value);
        } catch (IllegalArgumentException e) {
            throw new TlvConvertException("unsigned int", e.getMessage());
        }
    }

    /**
     * Attempts to get the underlying value as a long.  If the underlying byte
     * array cannot be converted, then an exception is thrown.  For example,
     * if the underlying byte array isn't a length of 8, then it cannot be
     * converted to a long.
     * @return The byte array value as a long
     * @throws TlvConvertException Thrown if the underlying byte array cannot
     *      be converted to a long.
     */
    public long getValueAsLong() throws TlvConvertException {
        try {
            return ByteArrayUtil.toLong(this.value);
        } catch (IllegalArgumentException e) {
            throw new TlvConvertException("long", e.getMessage());
        }
    }

    /**
     * Checks if another Tlv object is equal to this Tlv object. Two tlv objects
     * are considered equal if they have the same tag and the same value.
     * 
     * @param t
     * @return true if this and t have the same tag and value, false otherwise.
     */
    @Override
    public boolean equals(Object t){
        if (t == null){
            return false;
        } else if(this == t){
            return true;
        }
        Tlv other = (Tlv) t;
        return other.tag == this.tag && Arrays.equals(other.value, this.value);
    }

    /**
     * Computes a valid hash code for this Tlv object.
     * 
     * @return int hash code for this object.
     */
    @Override
    public int hashCode(){
        int hash = 42;
        hash = 31 * hash + ((int) this.tag);
        hash = 31 * hash + Arrays.hashCode(this.value);
        return hash;
    }
}
